core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -924,25 +900,33 @@ private object RTS {
       io.ensuring(exitUninterruptible)
     }
 
-    final def register(cb: Callback[E, A]): Async[E, A] = join0(cb)
+    final def register(cb: Callback[E, A]): Async[E, A] =
+      observe0 {
+        case ExitResult.Completed(r) => cb(r)
+        case _                       =>
```

Comment:
Why is any other case than `Completed` ignored? This seems to imply the callback will not be invoked in the other cases.

Reply:
I think I answered my own question, but I think you should match against the other two, even though they're impossible. In the case of `ExitResult.Failed(v)`, `v` will have type `Nothing`, so you can return it as the result of `register` (it's dead code, of course). In the case of `Terminated`, I'm not sure what to do. In theory it's not reachable, but if it were reachable, it would indicate a defect in RTS and it would be nice to capture that somehow (maybe uncaught error handler?).